### PR TITLE
Issue #1632 external object issues

### PR DESCRIPTION
**Pull request by elegoff**
Connected to https://github.com/slipstream/SlipStreamServer/issues/1632

### CODE REVIEW

File: cimi-resources/src/com/sixsq/slipstream/ssclj/resources/external_object/utils.clj [hunk 1]
```diff
@@ -45,7 +51,109 @@
...
+      (check-authn-fn request)))
+
+
+(defn uploadable-bucket?
+  "When the bucket exists, but the user can't create the object.
+  The external object resource must not be created."
+  [obj-store-conf bucketName]
+  (let [upload-url (generate-url obj-store-conf bucketName "test-probe" :put {:ttl 3 :content-type "text/plain" :filename "test.txt"})]
+    (= 200 (http/put upload-url))))
+
```
Comment (loomis) [inline cimi-resources/src/com/sixsq/slipstream/ssclj/resources/external_object/utils.clj]: I don't think that this implementation will work in practice.  The test file is never deleted. The second time the user tries to add an object to the same bucket, this request will fail and cause the function to incorrectly return false.  A cleaner solution with fewer error corner cases would be to check the ACL of the bucket itself, provided that the same calls work at both Amazon and Exoscale.

File: cimi-resources/src/com/sixsq/slipstream/ssclj/resources/external_object/utils.clj [hunk 2]
```diff
@@ -45,7 +51,109 @@
...
+        s3client (get-s3-client obj-store-conf)]
+    (cond
+      ;; When the user requests to delete an object, but it no longer exists :
+      ;; The external object resource should be deleted normally.
+      (not (bucket-exists? s3client bucketName)) resource
+      ;; "When the user cannot access the bucket/object to delete it :
+      ;; the external object resource should not be deleted."
+      (not (authorized-bucket-operation? resource request a/can-modify?)) (logu/log-and-throw 403 (format "Deleting the bucket %s is not allowed" bucketName))
+      :all-ok resource)))
+
```
Comment (loomis) [inline cimi-resources/src/com/sixsq/slipstream/ssclj/resources/external_object/utils.clj]: Shouldn't this also check the case where the bucket exists but the object does not?

File: cimi-resources/src/com/sixsq/slipstream/ssclj/resources/external_object.clj [hunk 3]
```diff
@@ -409,9 +383,9 @@
           (fn [resource _] (:objectType resource)))
 
 (defmethod delete-subtype :default
```
Comment (loomis) [inline cimi-resources/src/com/sixsq/slipstream/ssclj/resources/external_object.clj]: I think that this implementation does not treat correctly the case where the S3 delete request throws an exception; at the very least an informative message should be added to the thrown exception before propagating this back up the stack. 

I also don't think that the logic is correct when the underlying bucket/object doesn't exist.  In this case, the flag will indicate that the external object can be deleted and then the code will actually try to delete the non-existent bucket/object, which will then throw an exception that causes the delete request to fail.

Also, I wonder whether there's any benefit is having this as a multi-method.  Neither of the existing implementations specialize the behavior and generally I wouldn't see where this is necessary.

File: cimi-resources/src/com/sixsq/slipstream/ssclj/resources/external_object/utils.clj (hunk 1 above)
Comment (elegoff) [reply cimi-resources/src/com/sixsq/slipstream/ssclj/resources/external_object/utils.clj]: https://docs.aws.amazon.com/AmazonS3/latest/dev/acl-using-java-sdk.html

Comment (elegoff) [reply cimi-resources/src/com/sixsq/slipstream/ssclj/resources/external_object/utils.clj]: Implementation changed in favor a of a HEAD request to the bucket. As stated by the Javadoc (https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/AmazonS3Client.html#headBucket-com.amazonaws.services.s3.model.HeadBucketRequest-) "This operation is useful to determine if a bucket exists and you have permission to access it."

File: cimi-resources/src/com/sixsq/slipstream/ssclj/resources/external_object.clj (hunk 3 above)
Comment (elegoff) [reply cimi-resources/src/com/sixsq/slipstream/ssclj/resources/external_object.clj]: When bucket does exists, the cimi resource deletion now succeeds. A 403 is issued when the bucket ACL does not allow it.

File: cimi-resources/src/com/sixsq/slipstream/ssclj/resources/external_object/utils.clj (hunk 2 above)
Comment (elegoff) [reply cimi-resources/src/com/sixsq/slipstream/ssclj/resources/external_object/utils.clj]: done

**Pull request merged**